### ISSUE TITLE
Get the attention matrix for visualisation

### ISSUE THREAD

**Issue by bikashg** [MaximumEntropy/Seq2Seq-PyTorch#3]
How can I get the attention matrix from the Seq2SeqFastAttention class? I want to put that matrix for visualisation.

**Comment by MaximumEntropy** [comment]
The attention matrix for that model is here [https://github.com/MaximumEntropy/Seq2Seq-PyTorch/blob/master/model.py#L1148](url). Note that I overwrite that variable in the very next line so you'd probably need to hack it to change variable names a little bit. Also alpha is **not normalized** so it isn't a conventional attention matrix, however you could try doing a softmax on alpha to make it one.

**Comment by bikashg** [comment]
Thank you for your response. I am still new to pytorch but that line 1148 you told me is located inside the forward() function which needs both input_src and input_trg. For test cases, I wouldn't have the input_trg. So, how could I get the predictions and attentions for my test cases?

**Comment by MaximumEntropy** [comment]
To decode, the initial input_trg will be just be the start token `<s>` for every item in the minibatch. `input_lines_trg = Variable(torch.LongTensor(
            [
                [trg['word2id']['<s>']]
                for i in xrange(batch_size)
            ]
        )).cuda()`

If you run forward once with your input_src and the above input_trg you'll get an un-normalized vector of vocab size. You can then softmax followed by argmax to get the most likely next words. You can concatenate these words to input_trg and then run forward again and so on. You can refer to evaluate.py to see how I do this - https://github.com/MaximumEntropy/Seq2Seq-PyTorch/blob/master/evaluate.py#L140

**Comment by bikashg** [comment]
Thank you for the info. It was really helpful. Just in case, I want to get attentions from the classes Seq2SeqAttention and Seq2SeqAttentionSharedEmbedding, what lines should I be looking it? Also, are they more accurate than those proposed by Seq2SeqFastAttention?

**Comment by MaximumEntropy** [comment]
In the Seq2SeqAttention model, you can use the "alpha" defined in this line https://github.com/MaximumEntropy/Seq2Seq-PyTorch/blob/master/model.py#L371 and return that.
